### PR TITLE
refactor(core): new back&forward button base on workbench

### PR DESCRIPTION
**Pull request by EYHN**
# feature:

## In Browser:
- hidden back&forward button in sidebar.
- back and forward is equal with `window.history.back()` `window.history.forward()` 

## In Desktop:
- Back and forward can be controlled through the sidebar, cmdk, and shortcut keys.
- back and forward act on the currently **active** view.
- buttons change disable&enable style based on current active view history

# Refactor:

Move app-sidebar and app-container from @affine/component to @affine/core

### CODE REVIEW

**Comment by EYHN** [comment]
* **#6012** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6012?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EYHN and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/core/src/modules/navigation/entities/navigator.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+                entries: history.entries,
+              });
+            });
+          })
+      )
+    ),
+    { index: 0, entries: [] }
+  );
+
+  readonly backable = this.location.map(
```
Comment (pengx17) [inline packages/frontend/core/src/modules/navigation/entities/navigator.ts:32]: does this work for the browser?

File: packages/frontend/core/src/utils/navigable-history.ts [hunk 2]
```diff
@@ -0,0 +1,221 @@
+import type { Blocker, Listener, Location, To } from 'history';
+import {
+  Action,
+  createPath,
+  type MemoryHistory,
+  type MemoryHistoryOptions,
+  parsePath,
+} from 'history';
+
+export interface NavigableHistory extends MemoryHistory {
```
Comment (forehalo) [inline packages/frontend/core/src/utils/navigable-history.ts:10]: what's the difference from react-router's memory history

Comment (EYHN) [reply packages/frontend/core/src/utils/navigable-history.ts:10]: react-router v6 only has MemoryRouter, which internally uses MemoryHistory of 'history' package.

Comment (EYHN) [reply packages/frontend/core/src/utils/navigable-history.ts:10]: MemoryRouter also includes some async loader and other functions, making it very complex and difficult to use.

File: packages/frontend/core/src/modules/navigation/entities/navigator.ts (hunk 1 above)
Comment (EYHN) [reply packages/frontend/core/src/modules/navigation/entities/navigator.ts:32]: The browser does not use this.
back&forward buttons are hidden on the browser

File: packages/frontend/core/src/components/app-sidebar/sidebar-header/index.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { SidebarSwitch } from './sidebar-switch';
+
+export const SidebarHeader = () => {
+  const open = useAtomValue(appSidebarOpenAtom);
+
+  return (
+    <div
+      className={navHeaderStyle}
+      data-open={open}
+      data-is-macos-electron={environment.isDesktop && environment.isMacOs}
```
Comment (darkskygit) [inline packages/frontend/core/src/components/app-sidebar/sidebar-header/index.tsx:15]: which place use this?

Comment (EYHN) [reply packages/frontend/core/src/components/app-sidebar/sidebar-header/index.tsx:15]: Used in css, leaving space for traffic lights on macos.

**Comment by pengx17** [comment]
It seems history back/forward with keyboard shortcut does not work well.

**Review by pengx17 — APPROVED** [review]

**Review by pengx17 — APPROVED** [review]

**Review by CatsJuice — APPROVED** [review]

**Pull request merged**